File: lms-FE/src/features/schools/schemas.ts
import { z } from "zod";

export const PrincipalSchema = z.object({
    id: z.number(),
    name: z.string(),
    email: z.string(),
});

export const SchoolSchema = z.object({
    id: z.number(),
    name: z.string(),
    subscription_start: z.string(),
    subscription_end: z.string(),
    max_teachers: z.number(),
    created_at: z.string(),
    updated_at: z.string(),
    principal: PrincipalSchema.optional(),
});

export const SchoolCreateSchema = z.object({
    name: z.string().min(2, "Name must be at least 2 characters"),
    subscription_start: z.string().nonempty("Subscription start date is required"),
    subscription_end: z.string().nonempty("Subscription end date is required"),
    max_teachers: z.number().min(1, "Must allow at least 1 teacher"),
});

export const SchoolUpdateSchema = z.object({
    name: z.string().optional(),
    subscription_end: z.string().optional(),
    max_teachers: z.number().optional(),
});

export const AssignPrincipalSchema = z.object({
    user_id: z.number().min(1, "Valid User ID is required to assign principal"),
});

export type School = z.infer<typeof SchoolSchema>;
export type SchoolCreatePayload = z.infer<typeof SchoolCreateSchema>;
export type SchoolUpdatePayload = z.infer<typeof SchoolUpdateSchema>;
export type AssignPrincipalPayload = z.infer<typeof AssignPrincipalSchema>;

export interface PaginatedSchools {
    items: School[];
    total: number;
    page: number;
    size: number;
    pages: number;
}

File: lms-BE/app/features/schools/schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional

class SchoolBase(BaseModel):
    name: str
    subscription_end: datetime
    max_teachers: int = 10

class PrincipalInfo(BaseModel):
    id: int
    name: str
    email: str

class SchoolCreate(SchoolBase):
    subscription_start: Optional[datetime] = None

class SchoolUpdate(BaseModel):
    name: Optional[str] = None
    subscription_end: Optional[datetime] = None
    max_teachers: Optional[int] = None

class SchoolRead(SchoolBase):
    id: int
    subscription_start: datetime
    created_at: datetime
    updated_at: datetime
    principal: Optional[PrincipalInfo] = None

    model_config = ConfigDict(from_attributes=True)

class SchoolPagination(BaseModel):
    items: list[SchoolRead]
    total: int
    page: int
    size: int
    pages: int

class AssignPrincipal(BaseModel):
    user_id: int
